docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst

```diff
@@ -0,0 +1,421 @@
+
+.. code-block:: cpp
+
+  struct A {
+    const int *p;
+    int x;
+  };
+
+  struct B {
+    A *a;
+    int y;
+  };
+
+  void sink(int);
+  int user_input();
+  void write_user_input_to(void *);
+  void read_data(const void *);
+
+A regular dataflow query such as the following query
+
+.. code-block:: ql
+
+  /**
+  * @kind path-problem
+  */
+
+  import semmle.code.cpp.dataflow.new.DataFlow
+  import Flow::PathGraph
+
+  module Config implements DataFlow::ConfigSig {
+    predicate isSource(DataFlow::Node source) {
+      source.asExpr().(Call).getTarget().hasName("user_input")
+    }
+
+    predicate isSink(DataFlow::Node sink) {
+      exists(Call call |
+        call.getTarget().hasName("sink") and
+        sink.asExpr() = call.getAnArgument()
+      )
+    }
+  }
+
+  module Flow = DataFlow::Global<Config>;
+
+  from Flow::PathNode source, Flow::PathNode sink
+  where Flow::flowPath(source, sink)
+  select sink.getNode(), source, sink, "Flow from user input to sink!"
+
+will catch most things á la
+
+.. code-block:: cpp
+
+  void fill_structure(B* b, const int* pu) {
+    // ...
+    b->a->p = pu;
+  }
+
+  void process_structure(const B* b) {
+    sink(*b->a->p);
+  }
+
+  void get_and_process() {
+    int u = user_input();
+    B* b = (B*)malloc(sizeof(B));
+    // ...
+    fill_structure(b, &u);
+    // ...
+    process_structure(b);
+    free(b);
+  }
+
+This is simple to match because we see:
+  1. User input starts at ``user_input()`` and flows into ``fill_structure``.
+  2. The data is written to the object ``b`` with access path ``[a, p]``.
+  3. The object ``b`` flows out of ``fill_structure`` and into ``process_structure``.
+  4. The access path ``[a, p]`` is read in ``process_structure`` and the value ends up in the sink.
+
+Flow from a qualifier to a field access
+---------------------------------------
+
+However, sometimes the writes or reads are not visible to CodeQL (for example, because the implementation of the function isn’t included in the database), and so dataflow won't be able to match up all stores with reads, and thus you don't get the result you want. For example, consider the following example:
+
+.. code-block:: cpp
+
+  void write_user_input_to(void*);
+  void use_value(int);
+  void* malloc(unsigned long);
+
+  struct U {
+    const int* p;
+    int x;
+  };
+
+  void process_user_data(const int* p) {
+    // ...
+    use_value(*p);
+  }
+
+  void get_and_process_user_input_v2() {
+    U* u = (U*)malloc(sizeof(U));
+    write_user_input_to(u);
+    process_user_data(u->p);
+    free(u);
+  }
+
+Here, flow starts at the outgoing argument of ``write_user_input_to(...)`` and proceeds to ``u->p``. However, because CodeQL has not observed a write to p prior to the read ``u->p``, dataflow will stop at ``u``. In order to convince CodeQL to proceed we need to add an additional flow step through field reads like so:
```

Comment:
> Here, flow starts at the outgoing argument of ``write_user_input_to(...)``

I doesn't though, because you haven't introduced `write_user_input_to` into `isSource` yet.  I think we should consider moving the QL below that adds the source and sink (but perhaps not the additional flow step) to just above this paragraph?

Reply:
That's a good point. I've tried to reorganize things in d6ccd63ed2db16b4c47f298bf00d0bcb64c9142c so that the relevant `isSource` is mentioned first. Does that improve the readability?

Reply:
Yes, much better!